converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Observation/ObservationUpsertHandler.cs

```diff
@@ -0,0 +1,105 @@
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Hl7.Fhir.Model;
+using Hl7.Fhir.Utility;
+using Microsoft.Health.Dicom.Client.Models;
+using Microsoft.Health.DicomCast.Core.Extensions;
+using Microsoft.Health.DicomCast.Core.Features.Fhir;
+
+namespace Microsoft.Health.DicomCast.Core.Features.Worker.FhirTransaction
+{
+    /// <inheritdoc/>
+    class ObservationUpsertHandler : IObservationUpsertHandler
+    {
+        private readonly IFhirService _fhirService;
+
+        public ObservationUpsertHandler(IFhirService fhirService)
+        {
+            EnsureArg.IsNotNull(fhirService, nameof(fhirService));
+            _fhirService = fhirService;
+        }
+
+        public async Task<FhirTransactionRequestEntry> BuildAsync(FhirTransactionContext context, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(context, nameof(context));
+            EnsureArg.IsNotNull(context.ChangeFeedEntry, nameof(context.ChangeFeedEntry));
+            EnsureArg.IsNotNull(context.Request, nameof(context.Request));
+
+            IResourceId patientId = context.Request.Patient.ResourceId;
+            IResourceId imagingStudyId = context.Request.ImagingStudy.ResourceId;
+            ChangeFeedEntry changeFeedEntry = context.ChangeFeedEntry;
+
+            // Parse all observations out of the dataset
+            ParsedObservation parsedObservations = ObservationParser.CreateObservations(
+                changeFeedEntry.Metadata,
+                patientId.ToResourceReference(),
+                imagingStudyId.ToResourceReference()
+            );
+            Collection<Observation> doseSummaries = parsedObservations.DoseSummaries;
+            Collection<Observation> irradiationEvents = parsedObservations.IrradiationEvents;
+
+            if (!doseSummaries.Any())
+            {
+                return null;
+            }
+
+            // operate on the pre-condition that their _should_ only be one DoseSummary per ImagingStudy.
+            // If multiple observations are found; just use the first one.
+            Observation doseSummary = doseSummaries[0];
+
+            // Filter out existing observations
+            Identifier imagingStudyIdentifier = imagingStudyId.ToResourceReference().Identifier;
+            IEnumerable<Observation> existingDoseSummariesAsync = imagingStudyIdentifier != null
+                ? await _fhirService
+                    .RetrieveObservationsAsync(
+                        imagingStudyId.ToResourceReference().Identifier,
+                        cancellationToken)
+                : new List<Observation>();
+
+            var existingDoseSummaries =
+                existingDoseSummariesAsync.ToList();
+            bool isExisting = existingDoseSummaries.Any();
+
+            FhirTransactionRequestEntry transactionRequestEntry;
+            if (isExisting)
```

Comment:
nit: if isExisting is not being used anywhere else, this can be if(existingDoseSummaries.Any()) ?